test/support/warning_extractor.rb

```diff
@@ -0,0 +1,51 @@
+require 'set'
+
+module RegexpParserTest
+  class Warning
+    class Filter
+      def initialize(whitelist)
+        @whitelist = whitelist
+      end
+
+      def assert_expected_warnings_only
+        original = $stderr
+        $stderr  = Extractor.new(original, @whitelist)
+
+        yield
+
+        assert_no_warnings($stderr.warnings)
+      ensure
+        $stderr = original
+      end
+
+    private
+
+      def assert_no_warnings(warnings)
+        fail "Unexpected warnings: #{warnings.to_a.join("\n")}" if warnings.any?
```

Comment:
Having a literal string inside an interpolation feels a bit weird: What about:

```
fail "Unexpected warnings: %s" % warnings.to_a.join("\n") if warnings.any?
```

Or even:

```
fail "Unexpected warnings: " + warnings.to_a.join("\n") if warnings.any?
```

Theoretically you can also go for `#<<` instead `#+` but I dislike to mutate unless there is a semantic need to cause a (visible) side effect.